### PR TITLE
chore: Add LockAndExecute util

### PR DESCRIPTION
**Pull request by palpatim**
Per [@kneekey23's suggestion](https://github.com/aws-amplify/amplify-ios/pull/1345#discussion_r679629689) #1345, added a locking utility to encapsulate a bit of the boilerplate around synchronizing code with NSLocking instances.

*Description of changes:*

*Check points:*

- [X] All unit tests pass
- [X] All integration tests pass
- [X] PR title conforms to conventional commit style

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by palpatim** [comment]
Hold off on this--just had a thought about a much better API

**Comment by kneekey23** [comment]
> Hold off on this--just had a thought about a much better API

good hopefully I can steal it then :)

**Review by palpatim — COMMENTED** [review]
Self review to set checkpoint

**Comment by palpatim** [comment]
Made this utility an extension on NSLocking instead.

File: Amplify/Core/Support/Internal/NSLocking+Execute.swift [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+import Foundation
+
+/// - Warning: Although this has `public` access, it is intended for internal use
+/// and should not be used directly by host applications. The behaviors and names of
+/// this type may change without warning.
+extension NSLocking {
+    /// Execute `block` after obtaining a lock on `lock`.
+    ///
+    /// - Warning: Although this has `public` access, it is intended for internal use
```
Comment (kneekey23) [inline Amplify/Core/Support/Internal/NSLocking+Execute.swift:16]: do you have to make it public for some reason and thus need this warning?

File: Amplify/Core/Support/Internal/NSLocking+Execute.swift [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import Foundation
+
+/// - Warning: Although this has `public` access, it is intended for internal use
+/// and should not be used directly by host applications. The behaviors and names of
+/// this type may change without warning.
+extension NSLocking {
```
Comment (kneekey23) [inline Amplify/Core/Support/Internal/NSLocking+Execute.swift:13]: also im stealing this :)

File: AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/MutationSync/OutgoingMutationQueue/SyncMutationToCloudOperation.swift [hunk 3]
```diff
@@ -162,43 +164,44 @@ class SyncMutationToCloudOperation: AsynchronousOperation {
...
             }
+        }
+    }
 
+    /// Initiates a locking context
+    private func respond(
+        toCloudResult result: GraphQLOperation<MutationSync<AnyModel>>.OperationResult,
+        withAPIRequest apiRequest: GraphQLRequest<MutationSync<AnyModel>>
+    ) {
+        lock.execute {
```
Comment (kneekey23) [inline AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/MutationSync/OutgoingMutationQueue/SyncMutationToCloudOperation.swift:191]: this one really cleans up the code, so elegant

File: AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/MutationSync/OutgoingMutationQueue/SyncMutationToCloudOperation.swift [hunk 4]
```diff
@@ -241,22 +240,8 @@ class SyncMutationToCloudOperation: AsynchronousOperation {
...
-        lock.lock()
-        defer {
-            lock.unlock()
-        }
-
-        let shouldRetry = (authTypesIterator?.count ?? 0) > 0
-        return RequestRetryAdvice(shouldRetry: shouldRetry, retryInterval: .milliseconds(0))
-    }
-
+    /// - Warning: Must be invoked from a locking context
```
Comment (kneekey23) [inline AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/MutationSync/OutgoingMutationQueue/SyncMutationToCloudOperation.swift:243]: Nice!

**Review by kneekey23 — APPROVED** [review]
LGTM! very nice clean up of the code with this extension. Since I lack some context on some of the code tho, i'd hope someone else from Amplify could also give you ship it but everywhere you just did a simple replace with the extension looks clean and excellent. also nice use of `rethrows` too!

File: Amplify/Core/Support/Internal/NSLocking+Execute.swift (hunk 1 above)
Comment (palpatim) [reply Amplify/Core/Support/Internal/NSLocking+Execute.swift:16]: Yes, in order to use Amplify methods in the separate packages, we have to have public accessibility.

File: Amplify/Core/Support/Internal/NSLocking+Execute.swift (hunk 2 above)
Comment (palpatim) [reply Amplify/Core/Support/Internal/NSLocking+Execute.swift:13]: It's open-source. All we need is attribution. :P

File: Amplify/Core/Support/Internal/NSLocking+Execute.swift (hunk 1 above)
Comment (kneekey23) [reply Amplify/Core/Support/Internal/NSLocking+Execute.swift:16]: ah gotcha makes sense, I understand this well

**Review by diegocstn — APPROVED** [review]

**Pull request merged**